cli/src/user.rs

```diff
@@ -327,128 +328,84 @@ impl User {
+            .mls_group
             .borrow_mut()
-            .stage_commit(
-                &create_commit_results.commit,
-                &proposal_store,
-                &[],
-                &self.crypto,
-            )
-            .expect("error applying commit");
-        group
-            .core_group
+            .add_members(&self.crypto, &[joiner_key_package])
+        {
+            Ok((out_msgs, welcome)) => (out_msgs, welcome),
+            Err(e) => return Err(format!("Failed to add member to group - {}", e)),
+        };
+
+        // First, process the invitation on our end.
+        let unverified_message = group
+            .mls_group
+            .borrow_mut()
+            .parse_message(out_messages.clone().into(), &self.crypto)
+            .expect("Could not parse message.");
+
+        let processed_message = group
+            .mls_group
             .borrow_mut()
-            .merge_commit(staged_commit)
-            .map_err(|e| format!("{}", e))?;
+            .process_unverified_message(unverified_message, None, &self.crypto)
+            .expect("Could not process unverified message.");
+
+        if let ProcessedMessage::StagedCommitMessage(staged_commit) = processed_message {
+            group
+                .mls_group
+                .borrow_mut()
+                .merge_staged_commit(*staged_commit)
+                .expect("Could not merge StagedCommit");
+        } else {
+            unreachable!("Expected a StagedCommit.");
+        }
 
-        // Send Welcome to the client.
+        // Second, send Welcome to the joiner.
         log::trace!("Sending welcome");
         self.backend
-            .send_welcome(&welcome_msg)
-            .expect("Error sending unwrap message");
+            .send_welcome(&welcome)
+            .expect("Error sending Welcome message");
 
-        // Send proposal to the group.
+        // Finally, send the MlsMessages to the group.
         log::trace!("Sending proposal");
         let group = groups.get_mut(group_id).unwrap(); // XXX: not cool.
         let group_recipients = self.recipients(group);
-        // TODO: the outgoing messages should use `MlsMessage` instead.
-        let msg = GroupMessage::new(
-            DsMlsMessage::Plaintext(VerifiableMlsPlaintext::from_plaintext(add_proposal, None)),
-            &group_recipients,
-        );
-        self.backend.send_msg(&msg)?;
-
-        // Send commit to the group.
-        log::trace!("Sending commit");
-        let msg = GroupMessage::new(
-            DsMlsMessage::Plaintext(VerifiableMlsPlaintext::from_plaintext(
-                create_commit_results.commit,
-                None,
-            )),
-            &group_recipients,
-        );
+        // XXX: Should we support MlsPlaintext messages?
+        let out_ds_messages = match out_messages {
+            MlsMessageOut::Ciphertext(msg) => DsMlsMessage::Ciphertext(*msg),
+            MlsMessageOut::Plaintext(_msg) => todo!(),
```

Comment:
If we don't support it, you could use `unimplemented!("We don't support plaintext messages.")`